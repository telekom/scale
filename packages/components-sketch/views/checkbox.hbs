{{!--
Scale https://github.com/telekom/scale

Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
--}}

{{#* inline "icon"}}M21.15,4.85 L9,17.05 L3.3,11.3 C2.9,10.9 2.3,10.9 1.9,11.3
C1.5,11.7 1.5,12.3 1.9,12.7 L9,19.85 L22.6,6.25 C23,5.85 23,5.25 22.6,4.85
C22.2,4.45 21.55,4.5 21.15,4.85 Z{{/inline}}

<div data-sketch-artboard="Checkbox" class="artboard">
  <h1 class="title--artboard">Checkbox</h1>

  <h2 class="title--artboard-section">Standard</h2>
  <div class="agent-states--grid">
    <div class="agent-states--item">
      <p class="agent-states--label">:available</p>
      <scale-checkbox
        data-sketch-key="checkbox-1"
        label="Checkbox Label"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:hover</p>
      <scale-checkbox
        data-sketch-key="checkbox-2"
        label="Checkbox Label"
        data-sketch-state="hover"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:focus</p>
      <scale-checkbox
        data-sketch-key="checkbox-3"
        label="Checkbox Label"
        data-sketch-state="focus"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:pressed</p>
      <scale-checkbox
        data-sketch-key="checkbox-4"
        label="Checkbox Label"
        data-sketch-state="active"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:disabled</p>
      <scale-checkbox
        data-sketch-key="checkbox-5"
        label="Checkbox Label"
        disabled
      ></scale-checkbox>
    </div>
  </div>

  <h2 class="title--artboard-section">Selected</h2>
  <div class="agent-states--grid">
    <div class="agent-states--item">
      <p class="agent-states--label">:available</p>
      <scale-checkbox
        checked
        data-sketch-key="checkbox-6"
        label="Checkbox Label"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:hover</p>
      <scale-checkbox
        checked
        data-sketch-key="checkbox-7"
        label="Checkbox Label"
        data-sketch-state="hover"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:focus</p>
      <scale-checkbox
        checked
        data-sketch-key="checkbox-8"
        label="Checkbox Label"
        data-sketch-state="focus"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:pressed</p>
      <scale-checkbox
        checked
        data-sketch-key="checkbox-9"
        label="Checkbox Label"
        data-sketch-state="active"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:disabled</p>
      <scale-checkbox
        checked
        data-sketch-key="checkbox-10"
        label="Checkbox Label"
        disabled
      ></scale-checkbox>
    </div>
  </div>

  <h2 class="title--artboard-section">Standard Info</h2>
  <div class="agent-states--grid">
    <div class="agent-states--item">
      <p class="agent-states--label">:available</p>
      <scale-checkbox
        data-sketch-key="checkbox-19"
        label="Checkbox Label"
        helper-text="Info message"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:hover</p>
      <scale-checkbox
        data-sketch-key="checkbox-20"
        label="Checkbox Label"
        helper-text="Info message"
        data-sketch-state="hover"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:focus</p>
      <scale-checkbox
        data-sketch-key="checkbox-21"
        label="Checkbox Label"
        helper-text="Info message"
        data-sketch-state="focus"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:pressed</p>
      <scale-checkbox
        data-sketch-key="checkbox-22"
        label="Checkbox Label"
        helper-text="Info message"
        data-sketch-state="active"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:disabled</p>
      <scale-checkbox
        data-sketch-key="checkbox-23"
        label="Checkbox Label"
        helper-text="Info message"
        disabled
      >
      </scale-checkbox>
    </div>
  </div>

  <h2 class="title--artboard-section">Selected Info</h2>
  <div class="agent-states--grid">
    <div class="agent-states--item">
      <p class="agent-states--label">:available</p>
      <scale-checkbox
        data-sketch-key="checkbox-24"
        checked
        label="Checkbox Label"
        helper-text="Info message"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:hover</p>
      <scale-checkbox
        data-sketch-key="checkbox-25"
        checked
        label="Checkbox Label"
        helper-text="Info message"
        data-sketch-state="hover"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:focus</p>
      <scale-checkbox
        data-sketch-key="checkbox-26"
        checked
        label="Checkbox Label"
        helper-text="Info message"
        data-sketch-state="focus"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:pressed</p>
      <scale-checkbox
        data-sketch-key="checkbox-27"
        checked
        label="Checkbox Label"
        helper-text="Info message"
        data-sketch-state="active"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:disabled</p>
      <scale-checkbox
        data-sketch-key="checkbox-28"
        checked
        label="Checkbox Label"
        helper-text="Info message"
        disabled
      >
      </scale-checkbox>
    </div>
  </div>

  <h2 class="title--artboard-section">Standard Error</h2>
  <div class="agent-states--grid">
    <div class="agent-states--item">
      <p class="agent-states--label">:available</p>
      <scale-checkbox
        status="error"
        helper-text="Error message"
        data-sketch-key="checkbox-11"
        label="Checkbox Label"
      >
      </scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:hover</p>
      <scale-checkbox
        status="error"
        helper-text="Error message"
        data-sketch-key="checkbox-12"
        label="Checkbox Label"
        data-sketch-state="hover"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:focus</p>
      <scale-checkbox
        status="error"
        helper-text="Error message"
        data-sketch-key="checkbox-13"
        label="Checkbox Label"
        data-sketch-state="focus"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:pressed</p>
      <scale-checkbox
        status="error"
        helper-text="Error message"
        data-sketch-key="checkbox-14"
        label="Checkbox Label"
        data-sketch-state="active"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:disabled</p>
    </div>
  </div>

  <h2 class="title--artboard-section">Selected Error</h2>
  <div class="agent-states--grid">
    <div class="agent-states--item">
      <p class="agent-states--label">:available</p>
      <scale-checkbox
        status="error"
        helper-text="Error message"
        checked
        data-sketch-key="checkbox-15"
        label="Checkbox Label"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:hover</p>
      <scale-checkbox
        status="error"
        helper-text="Error message"
        checked
        data-sketch-key="checkbox-16"
        label="Checkbox Label"
        data-sketch-state="hover"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:focus</p>
      <scale-checkbox
        status="error"
        helper-text="Error message"
        checked
        data-sketch-key="checkbox-17"
        label="Checkbox Label"
        data-sketch-state="focus"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:pressed</p>
      <scale-checkbox
        status="error"
        helper-text="Error message"
        checked
        data-sketch-key="checkbox-18"
        label="Checkbox Label"
        data-sketch-state="active"
      ></scale-checkbox>
    </div>
    <div class="agent-states--item">
      <p class="agent-states--label">:disabled</p>
    </div>
  </div>
</div>

<script src="/symbol_names.js"></script>
<script>
  Array.from(document.querySelectorAll('scale-checkbox')).forEach((btn) => {
    function evalNaming(btn) {
      if (
        btn.getAttribute('helper-text') &&
        btn.getAttribute('status') !== 'error'
      )
        return 'Info';
      if (btn.getAttribute('status') === 'error') return 'Error';
      return 'Standard';
    }
    btn.dataset.sketchSymbol = `Checkbox / ${order(
      1,
      btn.hasAttribute('checked') ? 'Selected' : 'Standard'
    )} / ${order(2, evalNaming(btn))} / ${order(4, getStateName(btn))}`;
  });
</script>
