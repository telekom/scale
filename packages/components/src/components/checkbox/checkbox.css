/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

scale-checkbox {
  --spacing-x: var(--scl-spacing-8);
  --transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  --color-text: var(--scl-color-text-standard);
  --color-error: var(--scl-color-text-error);
  --color-disabled: var(--scl-color-background-disabled);
  --color-standard: var(--scl-color-background-standard);
  --background-disabled: var(--scl-color-white);
  --color-primary: var(--scl-color-primary, #e20074);
  --color-focus: var(--scl-color-focus);
  --color-primary-hover: var(--scl-color-primary-hover, #f90984);
  --color-primary-active: var(--scl-color-primary-active, #cb0068);

  /*control*/
  --width-control: var(--scl-spacing-16);
  --height-control: var(--scl-spacing-16);
  --transition-control: var(--transition);
  --spacing-control: var(--scl-spacing-2);
  --spacing-left-control: var(--scl-spacing-4);
  --radius-control: var(--scl-radius-4);
  --border-width-control: var(--scl-spacing-1);

  /*helper-text*/
  --transition-helper-text: var(--transition);
  --spacing-left-helper-text: calc(var(--width-control) + var(--spacing-x));
  --spacing-top-helper-text: var(--scl-spacing-4);
  --font-size-helper-text: var(--scl-font-size-12);
  --font-weight-helper-text: var(--scl-font-weight-bold);
  --line-height-helper-text: var(--scl-font-line-height-133);
  --color-helper-text: var(--scl-color-blue-70);
  /*label*/
  --font-weight-label: var(--scl-font-weight-medium);
  --transition-label: var(--transition);

  /*icon*/
  --color-icon-checked-disabled: var(--scl-color-grey-30);
  --color-icon-checked-active: var(--scl-color-white);
  --width-icon: var(--scl-spacing-12);
  --height-icon: var(--scl-spacing-12);
  --left-icon: 50%;
  --top-icon: 50%;
  --transform-icon: translate(-50%, -50%);
}

.checkbox {
  position: relative;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  cursor: pointer;
}

.checkbox .checkbox__control-wrapper {
  width: var(--width-control);
  height: var(--height-control);
  display: flex;
  position: relative;
  align-items: center;
}

.checkbox__label-wrapper {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  cursor: pointer;
  gap: var(--spacing-x);
}

.checkbox .checkbox__label {
  color: var(--color-text);
  font-weight: var(--font-weight-label);
  transition: var(--transition-label);
  cursor: pointer;
}

.checkbox .checkbox__meta {
  width: 100%;
  display: flex;
  justify-content: space-between;
}

.checkbox .checkbox__helper-text {
  transition: var(--transition-helper-text);
  padding-left: var(--spacing-left-helper-text);
  font-size: var(--font-size-helper-text);
  font-weight: var(--font-weight-helper-text);
  line-height: var(--line-height-helper-text);
  color: var(--color-helper-text);
  margin-top: var(--spacing-top-helper-text);
}

.checkbox--status-error .checkbox__helper-text {
  color: var(--color-error);
}

.checkbox input {
  width: 0;
  height: 0;
  opacity: 0;
  position: absolute;
}

/* same as .checkbox--disabled .checkbox__label, (see rule further downwards)
  .checkbox input:disabled ~ .checkbox__label {
  color: var(--color-disabled);
} */

.checkbox--disabled.checkbox--checked .checkbox__label {
  color: var(--color-disabled);
}
.checkbox--disabled.checkbox--checked
  .checkbox__control-wrapper
  .checkbox__control {
  background: var(--color-disabled);
}
.checkbox--disabled.checkbox--indeterminate
  .checkbox__control-wrapper
  .checkbox__control {
  background: var(--color-disabled);
}
.checkbox--disabled.checkbox--checked
  .checkbox__control-wrapper
  .checkbox__icon {
  color: var(--color-icon-checked-disabled);
}
.checkbox--enabled.checkbox--checked
  input:hover
  ~ .checkbox__control-wrapper
  .checkbox__control/* ,
.checkbox
  input:checked:not([disabled])
  ~ .checkbox__control-wrapper:hover
  .checkbox__control  */ {
  box-shadow: none;
  border-color: var(--color-primary-hover);
  background: var(--color-primary-hover);
}
.checkbox--enabled.checkbox--indeterminate
  input:hover
  ~ .checkbox__control-wrapper
  .checkbox__control {
  box-shadow: none;
  border-color: var(--color-primary-hover);
  background: var(--color-primary-hover);
}
.checkbox--checked.checkbox--eanabled
  input:active
  ~ .checkbox__control-wrapper
  .checkbox__control/* ,
.checkbox
  input:checked:not([disabled])
  ~ .checkbox__control-wrapper:active
  .checkbox__control  */ {
  border-color: var(--color-primary-active);
  background: var(--color-primary-active);
}
.checkbox--indeterminate.checkbox--eanabled
  input:active
  ~ .checkbox__control-wrapper
  .checkbox__control/* ,
.checkbox
  input:checked:not([disabled])
  ~ .checkbox__control-wrapper:active
  .checkbox__control  */ {
  border-color: var(--color-primary-active);
  background: var(--color-primary-active);
}
.checkbox--enabled.checkbox--checked
  input:active
  ~ .checkbox__control-wrapper
  .checkbox__icon/* ,
.checkbox
  input:checked:not([disabled])
  ~ .checkbox__control-wrapper:active
  .checkbox__icon */ {
  color: var(--color-icon-checked-active);
}

.checkbox--enabled.checkbox--checked
  .checkbox__control-wrapper
  .checkbox__control {
  border: var(--scl-spacing-1) solid var(--color-primary);
  background: var(--color-primary);
}
.checkbox--enabled.checkbox--indeterminate
  .checkbox__control-wrapper
  .checkbox__control {
  border: var(--scl-spacing-1) solid var(--color-primary);
  background: var(--color-primary);
}
.checkbox.checkbox--disabled .checkbox__control-wrapper .checkbox__control {
  border-color: var(--color-disabled);
}
.checkbox input:focus ~ .checkbox__control-wrapper .checkbox__control {
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--color-focus);
}
.checkbox--enabled
  input:hover
  ~ .checkbox__control-wrapper
  .checkbox__control/* ,
.checkbox
  input:not([disabled])
  ~ .checkbox__control-wrapper:hover
  .checkbox__control */ {
  box-shadow: none;
  border-color: var(--color-primary-hover);
}
.checkbox--enabled
  input:hover
  ~ .checkbox__control-wrapper
  ~ .checkbox__label/* ,
.checkbox
  input:not([disabled])
  ~ .checkbox__control-wrapper:hover
  ~ .checkbox__label  */ {
  color: var(--color-primary-hover);
}
.checkbox.checkbox--enabled
  input:active
  ~ .checkbox__control-wrapper
  .checkbox__control/* ,
.checkbox
  input:not([disabled])
  ~ .checkbox__control-wrapper:active
  .checkbox__control  */ {
  border-color: var(--color-primary-active);
  background: var(--color-primary-active);
}

/* .checkbox input:not([disabled]):active ~ .checkbox__control-wrapper .checkbox__icon,
.checkbox input:not([disabled]) ~ .checkbox__control-wrapper:active .checkbox__icon {
  color: var(--color-primary-active);
} */

.checkbox.checkbox--enabled
  input:active
  ~ .checkbox__control-wrapper
  ~ .checkbox__label/* ,
.checkbox
  input:not([disabled])
  ~ .checkbox__control-wrapper:active
  ~ .checkbox__label */ {
  color: var(--color-primary-active);
}
.checkbox .checkbox__control-wrapper .checkbox__control {
  width: var(--width-control);
  height: var(--height-control);
  box-sizing: border-box;
  transition: var(--transition-control);
  border-radius: var(--radius-control);
  border: var(--border-width-control) solid var(--color-text);
  background: var(--color-standard);
}
.checkbox .checkbox__control-wrapper .checkbox__icon {
  top: var(--top-icon);
  left: var(--left-icon);
  width: var(--width-icon);
  height: var(--height-icon);
  transform: var(--transform-icon);
  position: absolute;
  user-select: none;
  color: var(--color-standard);
}
.checkbox .checkbox__control-wrapper .checkbox__icon svg {
  width: var(--width-icon);
  height: var(--height-icon);
}

.checkbox.checkbox--status-error .checkbox__control-wrapper .checkbox__control {
  border: var(--scl-spacing-2) solid var(--scl-color-background-error);
}

.checkbox--disabled .checkbox__label,
.checkbox--disabled input,
.checkbox--disabled .checkbox__control-wrapper,
.checkbox--disabled .checkbox__helper-text {
  cursor: not-allowed;
  border-color: var(--color-disabled);
  color: var(--color-disabled);
}
