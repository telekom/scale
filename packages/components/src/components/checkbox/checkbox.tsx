/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Prop,
} from '@stencil/core';
import { emitEvent } from '../../utils/utils';
import statusNote from '../../utils/status-note';

export interface CheckboxInterface extends HTMLElement {
  checked: boolean;
  indeterminate: boolean;
  disabled: boolean;
  value: string;
  label: string;
  ariaLabelCheckbox?: string;
}

let i = 0;
@Component({
  tag: 'scale-checkbox',
  styleUrl: './checkbox.css',
  shadow: false,
})
export class Checkbox {
  @Element() host: HTMLElement;
  /** (optional) Input name */
  @Prop() name?: string;
  /** (optional) Input label */
  @Prop() label: string = '';
  /** (optional) Input label output */
  @Prop() ariaLabelCheckbox?: string;
  /** (optional) Hides the specified label visually */
  @Prop() hideLabel?: boolean = false;
  /** (optional) Input helper text */
  @Prop() helperText?: string;
  /** @deprecated - invalid should replace status */
  @Prop() status?: string = '';
  /** (optional) Input status */
  @Prop() invalid?: boolean = false;
  /** (optional) Input disabled */
  @Prop({ reflect: true }) disabled?: boolean = false;
  /** (optional) Active switch */
  @Prop({ mutable: true, reflect: true }) checked?: boolean = false;
  /** (optional) indeterminate */
  @Prop({ mutable: true, reflect: true }) indeterminate?: boolean = false;
  /** (optional) Input value */
  @Prop() value?: string = '';
  /** (optional) Input checkbox id */
  @Prop({ mutable: true }) inputId?: string;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  /** Emitted when the value has changed. */
  @Event({ eventName: 'scale-change' }) scaleChange: EventEmitter;
  /** @deprecated in v3 in favor of kebab-case event names */
  @Event({ eventName: 'scaleChange' }) scaleChangeLegacy: EventEmitter;

  private id = i++;

  componentDidRender() {
    if (this.status !== '') {
      statusNote({
        tag: 'deprecated',
        message:
          'Property "status" is deprecated. Please use the "invalid" property!',
        type: 'warn',
        source: this.host,
      });
    }
    if (this.host.hasAttribute('aria-label')) {
      statusNote({
        tag: 'deprecated',
        message:
          'Property "ariaLabel" is deprecated. Please use the "ariaLabelCheckbox" property!',
        type: 'warn',
        source: this.host,
      });
    }
  }

  handleChange = (ev) => {
    if (this.indeterminate) {
      this.indeterminate = false;
      this.checked = true;
      ev.target.checked = true;
    } else {
      this.checked = ev.target.checked;
    }

    const { checked, indeterminate, value, disabled } = this;

    emitEvent(this, 'scaleChange', { checked, indeterminate, value, disabled });
  };

  connectedCallback() {
    if (!this.inputId) {
      this.inputId = 'input-checkbox-' + this.id;
    }
  }

  /* Accessibility: rendering the icon *only* when checked, otherwise is always visible in HCM */
  renderIcon() {
    if (this.indeterminate) {
      return (
        <scale-icon-action-indeterminate
          part="icon"
          decorative
        ></scale-icon-action-indeterminate>
      );
    }

    if (this.checked) {
      return (
        <scale-icon-action-success
          part="icon"
          decorative
        ></scale-icon-action-success>
      );
    }
  }

  renderHelperText(text) {
    if (this.helperText && this.helperText !== '') {
      return (
        <div
          part="helper-text"
          id={text.id}
          aria-live="polite"
          aria-relevant="additions removals"
        >
          {text.content}
        </div>
      );
    }
  }

  render() {
    const helperText = {
      id: this.helperText ? `helper-text-${this.id}` : null,
      content: this.helperText,
    };

    return (
      <Host
        class={{
          checked: this.checked,
          indeterminate: this.indeterminate,
          disabled: this.disabled,
          error: this.status === 'error' || this.invalid,
          hideLabel: this.hideLabel,
        }}
      >
        <input
          type="checkbox"
          part="input"
          name={this.name || null}
          id={this.inputId}
          value={this.value}
          checked={this.checked}
          indeterminate={this.indeterminate}
          aria-label={this.ariaLabelCheckbox}
          aria-checked={this.indeterminate ? 'mixed' : false}
          aria-invalid={this.status === 'error' || this.invalid}
          aria-describedBy={helperText.id}
          disabled={this.disabled}
          onChange={this.handleChange}
        />
        <label part="container" htmlFor={this.inputId}>
          <div part="checkbox">{this.renderIcon()}</div>
          {/* TODO: discuss deprecation of the slot (move closer so W3C spec) */}
          <div part="label">{this.label || <slot></slot>}</div>
        </label>
        {this.renderHelperText(helperText)}
      </Host>
    );
  }
}
