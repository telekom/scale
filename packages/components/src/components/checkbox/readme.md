# scale-checkbox



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute       | Description                    | Type               | Default     |
| --------------- | --------------- | ------------------------------ | ------------------ | ----------- |
| `checked`       | `checked`       | (optional) Active switch       | `boolean`          | `false`     |
| `disabled`      | `disabled`      | (optional) Input disabled      | `boolean`          | `undefined` |
| `helperText`    | `helper-text`   | (optional) Input helper text   | `string`           | `''`        |
| `indeterminate` | `indeterminate` | (optional) indeterminate       | `boolean`          | `false`     |
| `inputId`       | `input-id`      | (optional) Input checkbox id   | `string`           | `undefined` |
| `label`         | `label`         | (optional) Input label         | `string`           | `''`        |
| `name`          | `name`          | (optional) Input name          | `string`           | `''`        |
| `status`        | `status`        | (optional) Input status        | `string`           | `''`        |
| `styles`        | `styles`        | (optional) Injected CSS styles | `string`           | `undefined` |
| `value`         | `value`         | (optional) Input value         | `number \| string` | `''`        |


## Events

| Event         | Description                         | Type               |
| ------------- | ----------------------------------- | ------------------ |
| `scaleChange` | Emitted when the value has changed. | `CustomEvent<any>` |


## Dependencies

### Used by

 - [scale-data-grid](../data-grid)

### Depends on

- [scale-icon-action-success](../icons/action-success)

### Graph
```mermaid
graph TD;
  scale-checkbox --> scale-icon-action-success
  scale-data-grid --> scale-checkbox
  style scale-checkbox fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
