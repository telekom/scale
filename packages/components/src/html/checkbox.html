<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <title>Stencil Component Starter</title>

    <script type="module" src="/build/scale-components.esm.js"></script>
    <link rel="stylesheet" href="/build/scale-components.css" />
    <style type="text/css" media="screen, print">
      html {
        margin: 0;
        padding: 0;
      }

      body {
        margin: 0;
        padding: 4rem;
      }
    </style>
  </head>

  <body>
    <p>
      <scale-checkbox
        checked
        disabled
        label="Selected Disabled"
      ></scale-checkbox>
    </p>
    <p>
      <scale-checkbox indeterminate label="Indeterminate"></scale-checkbox>
    </p>
    <p>
      <scale-checkbox
        label="Standard with error"
        helper-text="This is the error message"
        invalid
      ></scale-checkbox>
    </p>
    <p>
      <scale-switch
        label="My styles were affecting checkbox styles"
      ></scale-switch>
    </p>
  </body>
</html>
