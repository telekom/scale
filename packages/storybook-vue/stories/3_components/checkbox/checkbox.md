<div style="display: inline-flex; align-items: center; justify-content: space-between; width: 100%;">
    <h1>Checkbox</h1>
    <img src="assets/aa.png" alt="Accessible AA" />
</div>

With Checkboxes, users toggle options on or off. They can answer simple yes-no questions in this way. Furthermore, users can select one or more options from a number of choices.

## General

A checkbox is a selection control that allows users to choose between two states. It consists of a control box and a corresponding label. Activating a checkbox does not affect other selections. It is considered to be selected when the control box has a checkmark. In the initial state, a checkbox can switch on or off.

![Image Name](assets/3_components/checkbox/checkbox.png)

### When to use

There are two main use cases for checkboxes:

- Single: Use a stand-alone checkbox if users are to turn an option on or off. For instance, this could be a request in a form or the agreement of general terms and conditions before buying a product.
- Multiple: Use a number of checkboxes if users are to select one or more options from a list. For instance, this could be a filter panel in a shop, a collection of settings, or a list of items to edit.

### When not to use

- When states are mutually exclusive, use radio buttons instead.

## Checkbox label

The label explicitly describes what happens when users select or deselect a checkbox. It should convey two opposite states. Always formulate the label positively to avoid double negatives.

![Image Name](assets/3_components/checkbox/checkbox_label.png)

## Disabled state

Disable the checkbox if:

- another action has to take place before the checkbox is usable or
- the option needs to be listed but is currently unavailable.

![Image Name](assets/3_components/checkbox/checkbox_disabled.png)

## Nested groups of checkboxes

> Available soon

When checkboxes are related to a parent checkbox, you can nest them. Selecting the parent checkbox automatically selects all checkboxes of the group and vice versa. When the user selects a portion of the nested checkboxes, the state of the parent checkbox is indeterminate.

![Image Name](assets/3_components/checkbox/checkbox_nesting.png)

## Required input

When checking a box is required, use an asterisk (\*) in the text label.

![Image Name](assets/3_components/checkbox/checkbox_required_input.png)

## Related components

<a href="?path=/usage/components-dropdown--standard">Dropdown, </a>
<a href="?path=/usage/components-switch--standard">Switch, </a>
<a href="?path=/usage/components-radio-button--standard">Radio Button</a>
