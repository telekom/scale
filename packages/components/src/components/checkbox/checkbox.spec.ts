/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { Checkbox } from './checkbox';

describe('Checkbox', () => {
  let page: SpecPage;

  beforeEach(async () => {
    page = await newSpecPage({
      components: [Checkbox],
      html: `<scale-checkbox></scale-checkbox>`,
    });
  });

  it('should match snapshot', async () => {
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot', async () => {
    const specPage = await newSpecPage({
      components: [Checkbox],
      html: `<scale-checkbox helper-text="helpertext"></scale-checkbox>`,
    });
    expect(specPage.root).toMatchSnapshot();
  });

  it('should reflect attributes', async () => {
    page = await newSpecPage({
      components: [Checkbox],
      html: `
        <scale-checkbox 
          name="testname"
          label="testlabel"
          checked
          disabled
          status="error"
          value="testvalue">
        </scale-checkbox>`,
    });
    await page.waitForChanges();
    expect(page.rootInstance.name).toBe('testname');
    expect(page.rootInstance.label).toBe('testlabel');
    expect(page.rootInstance.disabled).toBe(true);
    expect(page.rootInstance.checked).toBe(true);
    expect(page.rootInstance.status).toBe('error');
    expect(page.rootInstance.value).toBe('testvalue');
  });

  it('should handle css classes', () => {
    const element = new Checkbox();
    element.checked = true;
    expect(element.getCssClassMap()).toContain('checkbox--checked');
    element.disabled = true;
    expect(element.getCssClassMap()).toContain('checkbox--disabled');
    element.status = 'error';
    expect(element.getCssClassMap()).toContain('checkbox--status-error');
  });

  it('should emit on change', async () => {
    const changeSpy = jest.fn();
    const changeSpyLegacy = jest.fn();
    page.doc.addEventListener('scale-change', changeSpy);
    page.doc.addEventListener('scaleChange', changeSpyLegacy);
    const element = page.root.querySelector('input');
    element.dispatchEvent(new Event('change'));
    await page.waitForChanges();
    expect(changeSpy).toHaveBeenCalled();
    expect(changeSpyLegacy).toHaveBeenCalled();
  });

  it('should handle inputId with value null', async () => {
    page = await newSpecPage({
      components: [Checkbox],
      html: `
        <scale-checkbox 
         input-id
         >
        </scale-checkbox>`,
    });
    expect(page.rootInstance.inputId).toBe('');
  });
});
