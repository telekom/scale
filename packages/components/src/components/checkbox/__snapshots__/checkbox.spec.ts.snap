// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match snapshot 1`] = `
<scale-checkbox>
  <!---->
  <div class="checkbox checkbox--enabled checkbox--not-checked">
    <label class="checkbox__label-wrapper" htmlfor="input-checkbox-0">
      <div class="checkbox__control-wrapper">
        <span class="checkbox__control"></span>
      </div>
      <span class="checkbox__label"></span>
    </label>
    <input id="input-checkbox-0" type="checkbox" value="">
  </div>
</scale-checkbox>
`;

exports[`Checkbox should match snapshot 2`] = `
<scale-checkbox helper-text="helpertext">
  <!---->
  <div class="checkbox checkbox--enabled checkbox--not-checked">
    <label class="checkbox__label-wrapper" htmlfor="input-checkbox-2">
      <div class="checkbox__control-wrapper">
        <span class="checkbox__control"></span>
      </div>
      <span class="checkbox__label"></span>
      <div aria-live="polite" aria-relevant="additions removals" class="checkbox__meta" id="helper-message-3">
        <div class="checkbox__helper-text">
          helpertext
        </div>
      </div>
    </label>
    <input aria-describedby="helper-message-3" id="input-checkbox-2" type="checkbox" value="">
  </div>
</scale-checkbox>
`;
